src/scanner.cpp

```diff
@@ -0,0 +1,187 @@
+#include "scanner.hpp"
+
+
+#include "clox.hpp"
+#include "token_type.hpp"
+#include "util.hpp"
+#include <cctype>
+#include <charconv>
+#include <cmath>
+#include <print>
+#include <utility>
+
+Scanner::Scanner(std::string_view source) : source(source)
+{
+  std::println("Starting scanner on {}", source);
+  scan();
+}
+
+[[nodiscard]] bool Scanner::end_reached() const { return current >= source.length(); }
+
+char Scanner::advance()
+{
+  if (end_reached()) { return '\0'; }
+  return source.at(current++);
+}
+
+void Scanner::add_token(TokenType type) { add_token(type, {}); }
+
+void Scanner::add_token(TokenType type, LiteralType literal)
+{
+  auto lexeme = source.substr(start, current - start);
+  std::println("{}", lexeme);
+  tokens.push_back({ line, type, lexeme, std::move(literal) });
+}
+
+bool Scanner::match(char key)
+{
+  if (end_reached()) { return false; }
+
+  if (source.at(current) != key) { return false; }
+
+  ++current;
+  return true;
+}
+
+[[nodiscard]] char Scanner::peek() const
+{
+  if (end_reached()) { return '\0'; }
+  return source.at(current);
+}
+
+[[nodiscard]] char Scanner::peek_next() const
+{
+  if (current + 1 >= source.length()) { return '\0'; }
+  return source.at(current + 1);
+}
+
+
+void Scanner::scan_string()
+{
+  while (peek() != '"' && !end_reached()) {
+    if (peek() == '\n') { ++line; }
+    advance();
+  }
+
+  if (end_reached()) {
+    Clox::report(line, "Unterminated string");
+    return;
+  }
+
+  advance();
+
+  std::string value = source.substr(start + 1, current - start - 1);
+  add_token(TokenType::STRING, value);
+}
+
+void Scanner::scan_number()
+{
+  while (util::is_digit(peek())) { advance(); }
+  if (peek() == '.' && (util::is_digit(peek_next()))) {
+    advance();
+    while (util::is_digit(peek())) { advance(); }
+  }
+
+  float val = NAN;
+  std::string str_val = source.substr(start, current - start);
+  auto [ptr, ec] = std::from_chars(str_val.data(), str_val.data() + str_val.size(), val);
+  if (ec == std::errc()) { add_token(TokenType::NUMBER, val); }
+}
+
+void Scanner::scan_identifier()
+{
+  while (util::is_alphanumeric(peek())) { advance(); }
+  std::string lexeme = source.substr(start, current - start);
+  auto it = keywords.find(lexeme);
+  TokenType type = TokenType::IDENTIFIER;
+  if (it != keywords.end()) { type = it->second; }
```

Comment:
False case not handled

Reply:
It is implicit. 
If the keyword is not found it will remain `TokenType::IDENTIFIER`.